### DISCUSSION TITLE
Fix the problem of using the SD card with the screen at the same time

### DISCUSSION THREAD

**Discussion by salarizadi** [Moddable-OpenSource/moddable#1082] [Announcements]
Hello.

I ran into a problem with the SPI port when I wanted to use the display and the SD card at the same time, after searching the internet I couldn't find anything until I remembered to go to the display settings in my project manifest. I changed the SPI port from ILI9341 definitions to VSPI_HOST.

And yes, it worked. You can now view the help document I just updated :

https://github.com/salarizadi/moddable-sdk/blob/main/Modules/SD/README.md#notes

**Comment by phoddie** [comment]
Yes, you can change the port used by the ILI9341 display driver. But, generally it is better to keep it on HSPI because HSPI can run faster than VSPI. Instead, could you move the SD Card from HSPI to VSPI? You could even provide a manifest option to configure that, like the ILI9341 driver does.

**Reply by salarizadi** [reply]
Yes, it is absolutely true and now you can use and enjoy the new version :

https://github.com/Moddable-OpenSource/moddable/discussions/1083

**Comment by javascript-2020** [comment]
i was following the example from `https://github.com/Moddable-OpenSource/moddable/discussions/1082`
which led me to the following instructions `https://github.com/salarizadi/moddable-sdk/tree/main/Modules/SD`
i get the following error during compilation

```
In file included from C:\Espressif\frameworks\esp-idf-v5.4\components\fatfs\vfs/esp_vfs_fat.h:10,
                 from C:\Users\win11\projects\moddable\modules\files\sd\esp32\sdcardspi.c:13:
C:\Espressif\frameworks\esp-idf-v5.4\components\sdmmc\include/sd_protocol_types.h:31:10: fatal error: esp_dma_utils.h: No such file or directory
   31 | #include "esp_dma_utils.h"
      |          ^~~~~~~~~~~~~~~~~
compilation terminated.
```

i can see the first file at `C:\Espressif\frameworks\esp-idf-v5.4\components\fatfs\vfs` and thats being included
i can see the second file at `C:\Espressif\frameworks\esp-idf-v5.4\components\esp_hw_support\dma\include`

my c is quite rusty, am i missing some compiler include during the compilation by the moddable sdk

really im looking to use spi devices with the moddalbe sdk, but i cant find some examples to get me going

**Reply by salarizadi** [reply]
I created the SD Card module for ESP-IDF 4, and you won't be able to use it with ESP-IDF 5 because some of the includes are no longer available or need to be replaced with others. At the moment, I don’t have time to update the module for ESP-IDF 5. If you’re not in a hurry, you can wait until I release an updated version. But if you need it urgently, you're welcome to take a look at the code I wrote and create your own version compatible with ESP-IDF 5:

https://github.com/salarizadi/moddable-sdk/tree/main/Modules/SD

**Reply by javascript-2020** [reply]
i see that makes sense, i'll take a proper look over the next couple of days

**Reply by phoddie** [reply]
>my c is quite rusty, am i missing some compiler include during the compilation by the moddable sdk
>
> really im looking to use spi devices with the Moddable SDK, but i cant find some examples to get me going

Good news: you don't need to use C. You can use JavaScript.

Our preferred [SPI API](https://419.ecma-international.org/#-10-io-classes-serial-peripheral-interface-spi) is the IO Class Pattern from the ECMA-419 standard. You can find examples for using that in [$MODDABLE/examples/io/spi](https://github.com/Moddable-OpenSource/moddable/tree/public/examples/io/i2c). The SDK also includes several drivers that use the API, including ePaper drivers [gdew0154m09](https://github.com/Moddable-OpenSource/moddable/tree/public/modules/drivers/gdew0154m09) and [it8951](https://github.com/Moddable-OpenSource/moddable/tree/public/modules/drivers/it8951).

**Reply by javascript-2020** [reply]
@phoddie ive written my first moddable module, the sdcard is working, it probably still requires some testing, when i write a full blown application with it

https://github.com/javascript-2020/micro-controllers/tree/main/xs-js/code/sdcard/sdcard-test

i was wondering a couple of things

1.  is there a way to include a git directory rather than a full blown repository in the manifest file, if the code that processes the manifest is open source i could take a look at it

2  is there a way to develop esp-idf components locally without having to upload them to the esp reigstry https://components.espressif.com/components/javascript-2020/sdcard

**Reply by phoddie** [reply]
Congratulations on your progress.

On your questions:

1. The git import always imports the full repository, but you can reference manifests and source code anywhere within that. Please see the [documentation on including Git repositories](https://www.moddable.com/documentation/tools/manifest#include-git) for examples.
2. Regarding ESP Registry development, that mechanism is developed by Espressif, not Moddable. We are just users of it, like you. ;) Maybe Espressif has a technique for that?
